### PR TITLE
HTTP CONNECT proxy support

### LINKED ISSUE #309: [Feature Request] Support for HTTP CONNECT proxies
### Is your feature request related to a problem? Please describe.

gRPC implementations in JS, Go, and Java all support the `GRPC_PROXY`/`HTTP_PROXY` env vars from https://github.com/grpc/grpc/blob/master/doc/environment_variables.md which allow proxying via `HTTP CONNECT` based proxies, but I can find no such support in https://github.com/hyperium/tonic.

### Describe the solution you'd like

Ideally we could contrib support, but in the absence of that, we may need to call out somewhere that gRPC proxy environment variables are not supported for TypeScript workers and that Envoy or another separate proxy should be used.

This issue is just investigatory to confirm what we do and don't support here.

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

* Added new `client::proxy` module for HTTP CONNECT support
  * Hyper/Tonic don't support this natively, so some workarounds had to be done
  * Based on user need, this also includes authentication
* Added new `http_connect_proxy` client option

Can't really perform a test of this without a more involved integration test that has a full proxy server implementation. I have manually tested this with Python SDK using this version of core and connecting to Temporal cloud through the proxy. Now that proxies are officially supported by Temporal, there's a feature PR at https://github.com/temporalio/features/pull/447 for testing them, and Python will be added once this is merged and the Python SDK is updated

## Checklist

1. Closes #309

### CODE REVIEW

File: client/src/proxy.rs
Comment (Sushisource) [inline client/src/proxy.rs:18]: Definitely ought to say what the two string fields are

**Review by Sushisource — APPROVED** [review]
Just one docstring to fix. Looks like this was fairly straightforward, nice.

File: client/src/proxy.rs
Comment (cretz) [reply client/src/proxy.rs]: Fixed

**Comment by cretz** [comment]
> Looks like this was fairly straightforward, nice.

The result was fairly straightforward but this took a lot of time because I had to dig through the internals because Tonic/Hyper don't make this easy. They don't make it easy to change the target TCP stream and they don't make it easy to do a URL call + upgrade after TCP connection. I also had to setup proxies to test with and verify this behavior matches existing Go and Java gRPC behavior. But yes, luckily the amount of code needed is small.

(waiting on external CI fix before merging)

**Pull request merged**
